File: agenttel-python/src/agenttel/engine.py
"""AgentTel engine — high-level orchestrator combining all components."""

from __future__ import annotations

from pathlib import Path

from opentelemetry import trace
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider

from agenttel.anomaly.detector import AnomalyDetector
from agenttel.baseline.composite import CompositeBaselineProvider
from agenttel.baseline.provider import StaticBaselineProvider
from agenttel.baseline.rolling import RollingBaselineProvider
from agenttel.causality.tracker import CausalityTracker
from agenttel.config import AgentTelConfig
from agenttel.enums import BaselineSource
from agenttel.error.classifier import ErrorClassifier
from agenttel.events import AgentTelEventEmitter
from agenttel.models import OperationBaseline, SloDefinition
from agenttel.processor import AgentTelSpanProcessor
from agenttel.slo.tracker import SloTracker
from agenttel.topology.registry import TopologyRegistry


class AgentTelEngine:
    """High-level API that wires all AgentTel components together."""

    def __init__(self, config: AgentTelConfig) -> None:
        self._config = config

        # Initialize components
        self._topology = TopologyRegistry(config)
        self._static_baselines = StaticBaselineProvider()
        self._rolling_baselines = RollingBaselineProvider(
            window_size=config.baselines.rolling_window_size,
            min_samples=config.baselines.rolling_min_samples,
        )
        self._baseline_provider = CompositeBaselineProvider(
            [self._rolling_baselines, self._static_baselines]
        )
        self._anomaly_detector = AnomalyDetector(
            z_score_threshold=config.anomaly_detection.z_score_threshold
        )
        self._slo_tracker = SloTracker()
        self._error_classifier = ErrorClassifier()
        self._causality_tracker = CausalityTracker()
        self._event_emitter = AgentTelEventEmitter()

        # Register dependencies in causality tracker
        for dep in config.dependencies:
            self._causality_tracker.register_dependency(dep.name, dep.criticality)

        # Register static baselines from config
        self._register_static_baselines()

        # Register SLOs from config
        self._register_slos()

        # Create span processor
        self._processor = AgentTelSpanProcessor(
            config=config,
            topology=self._topology,
            baseline_provider=self._baseline_provider,
            rolling_provider=self._rolling_baselines,
            anomaly_detector=self._anomaly_detector,
            slo_tracker=self._slo_tracker,
            error_classifier=self._error_classifier,
            causality_tracker=self._causality_tracker,
            event_emitter=self._event_emitter,
        )

    @classmethod
    def from_config(cls, config_path: str | Path = "agenttel.yml") -> AgentTelEngine:
        """Create engine from a YAML config file."""
        config = AgentTelConfig.from_yaml(config_path)
        return cls(config)

    @property
    def config(self) -> AgentTelConfig:
        return self._config

    @property
    def processor(self) -> AgentTelSpanProcessor:
        return self._processor

    @property
    def topology(self) -> TopologyRegistry:
        return self._topology

    @property
    def baseline_provider(self) -> CompositeBaselineProvider:
        return self._baseline_provider

    @property
    def rolling_baselines(self) -> RollingBaselineProvider:
        return self._rolling_baselines

    @property
    def anomaly_detector(self) -> AnomalyDetector:
        return self._anomaly_detector

    @property
    def slo_tracker(self) -> SloTracker:
        return self._slo_tracker

    @property
    def error_classifier(self) -> ErrorClassifier:
        return self._error_classifier

    @property
    def causality_tracker(self) -> CausalityTracker:
        return self._causality_tracker

    @property
    def event_emitter(self) -> AgentTelEventEmitter:
        return self._event_emitter

    def install(self, tracer_provider: TracerProvider | None = None) -> None:
        """Install the AgentTel span processor into the OTel SDK."""
        if tracer_provider is None:
            # Create a new provider with topology resource attributes
            resource = Resource.create(self._topology.get_topology_attributes())
            tracer_provider = TracerProvider(resource=resource)
            trace.set_tracer_provider(tracer_provider)

        tracer_provider.add_span_processor(self._processor)

        # Emit deployment event if configured
        if self._config.deployment.emit_on_startup:
            self._event_emitter.emit_deployment(
                version=self._config.deployment.version,
                environment=self._config.deployment.environment,
                deployment_id=self._config.deployment.deployment_id,
            )

    def _register_static_baselines(self) -> None:
        """Register static baselines from operation configs."""
        for op_name, op_config in self._config.operations.items():
            p50 = op_config.latency_p50_ms
            p99 = op_config.latency_p99_ms
            if p50 is not None and p99 is not None:
                baseline = OperationBaseline(
                    operation_name=op_name,
                    latency_p50_ms=p50,
                    latency_p99_ms=p99,
                    error_rate=op_config.expected_error_rate or 0.0,
                    source=BaselineSource.STATIC,
                )
                self._static_baselines.register(baseline)

    def _register_slos(self) -> None:
        """Register SLO definitions from config."""
        for slo_name, slo_config in self._config.slo.items():
            self._slo_tracker.register(
                SloDefinition(
                    name=slo_config.name or slo_name,
                    target=slo_config.target,
                    type=slo_config.type,
                )
            )

File: agenttel-node/src/engine.ts
import type { AgentTelConfig } from './config.js';
import { defaultConfig } from './config.js';
import { AnomalyDetector } from './anomaly/detector.js';
import { StaticBaselineProvider, type BaselineProvider } from './baseline/provider.js';
import { RollingBaselineProvider } from './baseline/rolling.js';
import { CompositeBaselineProvider } from './baseline/composite.js';
import { ErrorClassifier } from './error/classifier.js';
import { SLOTracker } from './slo/tracker.js';
import { TopologyRegistry } from './topology/registry.js';
import { AgentTelEventEmitter } from './events/emitter.js';
import { AgentTelSpanProcessor } from './processor.js';
import type { OperationBaseline, SLODefinition } from './interfaces.js';

/** Top-level orchestrator for AgentTel SDK. */
export class AgentTelEngine {
  readonly topology: TopologyRegistry;
  readonly rollingProvider: RollingBaselineProvider;
  readonly baselineProvider: BaselineProvider;
  readonly anomalyDetector: AnomalyDetector;
  readonly sloTracker: SLOTracker;
  readonly errorClassifier: ErrorClassifier;
  readonly eventEmitter: AgentTelEventEmitter;
  readonly spanProcessor: AgentTelSpanProcessor;

  private constructor(config: AgentTelConfig) {
    // Topology
    this.topology = new TopologyRegistry();
    if (config.topology) {
      if (config.topology.team) this.topology.team = config.topology.team;
      if (config.topology.tier) this.topology.tier = config.topology.tier;
      if (config.topology.domain) this.topology.domain = config.topology.domain;
      if (config.topology.onCallChannel) this.topology.onCallChannel = config.topology.onCallChannel;
      if (config.topology.repoUrl) this.topology.repoUrl = config.topology.repoUrl;
    }
    if (config.dependencies) {
      for (const dep of config.dependencies) {
        this.topology.registerDependency({
          name: dep.name,
          type: dep.type ?? '',
          criticality: dep.criticality ?? '',
          protocol: dep.protocol,
          timeoutMs: dep.timeoutMs,
          circuitBreaker: dep.circuitBreaker,
          fallback: dep.fallback,
          healthEndpoint: dep.healthEndpoint,
        });
      }
    }
    if (config.consumers) {
      for (const c of config.consumers) {
        this.topology.registerConsumer({ name: c.name, pattern: c.pattern ?? 'sync', slaLatencyMs: c.slaLatencyMs ?? 0 });
      }
    }

    // Baselines
    this.rollingProvider = new RollingBaselineProvider(
      config.rollingWindowSize ?? 1000,
      config.rollingMinSamples ?? 10,
    );

    const staticBaselines: Record<string, OperationBaseline> = {};
    if (config.operations) {
      for (const op of config.operations) {
        staticBaselines[op.name] = {
          operationName: op.name,
          latencyP50Ms: op.baselineLatencyP50Ms ?? 0,
          latencyP99Ms: op.baselineLatencyP99Ms ?? 0,
          errorRate: op.baselineErrorRate ?? 0,
          source: 'static',
          updatedAt: new Date(),
        };
      }
    }
    const staticProvider = new StaticBaselineProvider(staticBaselines);
    this.baselineProvider = new CompositeBaselineProvider(staticProvider, this.rollingProvider);

    // Anomaly detection
    this.anomalyDetector = new AnomalyDetector(config.anomalyZScoreThreshold ?? 3.0);

    // SLO tracking
    this.sloTracker = new SLOTracker();
    if (config.slos) {
      for (const slo of config.slos) {
        this.sloTracker.register({
          name: slo.name,
          operationName: slo.operationName,
          type: slo.type ?? 'availability',
          target: slo.target,
          windowSeconds: slo.windowSeconds,
        });
      }
    }

    // Error classifier
    this.errorClassifier = new ErrorClassifier();

    // Events
    this.eventEmitter = new AgentTelEventEmitter();

    // Span processor
    this.spanProcessor = new AgentTelSpanProcessor({
      topology: this.topology,
      baselineProvider: this.baselineProvider,
      rollingProvider: this.rollingProvider,
      anomalyDetector: this.anomalyDetector,
      sloTracker: this.sloTracker,
      errorClassifier: this.errorClassifier,
      eventEmitter: this.eventEmitter,
    });
  }

  /** Create a new builder. */
  static builder(): AgentTelEngineBuilder {
    return new AgentTelEngineBuilder();
  }

  /** Create an engine with a config directly. */
  static fromConfig(config: AgentTelConfig): AgentTelEngine {
    return new AgentTelEngine(config);
  }
}

/** Builder for configuring and creating an AgentTelEngine. */
export class AgentTelEngineBuilder {
  private config: AgentTelConfig = defaultConfig();

  withConfig(config: AgentTelConfig): this {
    this.config = { ...this.config, ...config };
    return this;
  }

  withTeam(team: string): this {
    if (!this.config.topology) this.config.topology = {};
    this.config.topology.team = team;
    return this;
  }

  withTier(tier: string): this {
    if (!this.config.topology) this.config.topology = {};
    this.config.topology.tier = tier;
    return this;
  }

  withDomain(domain: string): this {
    if (!this.config.topology) this.config.topology = {};
    this.config.topology.domain = domain;
    return this;
  }

  withSLO(slo: SLODefinition): this {
    if (!this.config.slos) this.config.slos = [];
    this.config.slos.push(slo);
    return this;
  }

  withAnomalyThreshold(threshold: number): this {
    this.config.anomalyZScoreThreshold = threshold;
    return this;
  }

  build(): AgentTelEngine {
    return AgentTelEngine.fromConfig(this.config);
  }
}
